quiz_subgraphs_ts/player/src/player.ts

```diff
@@ -0,0 +1,114 @@
+import { ApolloServer } from "@apollo/server";
+import { expressMiddleware } from "@apollo/server/express4";
+import cors from "cors";
+import bodyParser from "body-parser";
+const { json } = bodyParser;
+import express from "express";
+import { createServer } from "http";
+import { ApolloServerPluginDrainHttpServer } from "@apollo/server/plugin/drainHttpServer";
+import { buildSubgraphSchema } from "@apollo/subgraph";
+import { WebSocketServer } from "ws";
+import { useServer } from "graphql-ws/lib/use/ws";
+import gql from "graphql-tag";
+import { readFileSync } from "fs";
+
+function uuid() {
+  return Math.random().toString(36).substring(2, 9);
+}
+
+interface Player {
+  id: string;
+  name: string;
+  quizId: string;
+}
+
+const PLAYERS: Record<string, Player> = {};
+
+const typeDefs = gql(readFileSync("./player.graphql", { encoding: "utf-8" }));
+
+function playersForAQuiz(quizId: string) {
+  return Object.values(PLAYERS).filter((player) => player.quizId === quizId);
+}
+
+const resolvers = {
+  Player: {
+    __resolveReference(reference: Player) {
+      return Object.values(PLAYERS).find(
+        (player) =>
+          player.id === reference.id && player.quizId === reference.quizId
+      );
+    },
+  },
+
+  Query: {
+    player(_: any, { playerId }: { playerId: string }) {
+      return PLAYERS[playerId];
+    },
+
+    playersForAQuiz(_: any, { quizId }: { quizId: string }) {
+      return playersForAQuiz(quizId);
+    },
+  },
+
+  Mutation: {
+    createPlayer(
+      _: any,
+      { userName, quizId }: { userName: string; quizId: string }
+    ) {
+      const player: Player = {
+        id: uuid(),
+        name: userName,
+        quizId,
+      };
+      PLAYERS[player.id] = player;
+    },
+  },
+
+  Subscription: {
+    playersForAQuiz: {
+      subscribe: async function* (_: any, { quizId }: { quizId: string }) {
+        let count = 0;
+        while (true) {
+          const quizPlayers = playersForAQuiz(quizId);
+          yield { playersForAQuiz: quizPlayers };
+          await new Promise((resolve) => setTimeout(resolve, 1000));
+          if (count === quizPlayers.length) count = 0;
+        }
+      },
+    },
+  },
+};
+
+const app = express();
+const httpServer = createServer(app);
+
+const schema = buildSubgraphSchema({ typeDefs, resolvers });
+const wsServer = new WebSocketServer({
+  server: httpServer,
+  path: "/",
+});
+const serverCleanup = useServer({ schema }, wsServer);
+
+const server = new ApolloServer({
+  schema,
+  plugins: [
+    ApolloServerPluginDrainHttpServer({ httpServer }),
+    {
+      async serverWillStart() {
+        return {
+          async drainServer() {
+            await serverCleanup.dispose();
+          },
+        };
+      },
+    },
+  ],
+});
+
+await server.start();
+app.use("/", cors(), json(), expressMiddleware(server));
+
+const PORT = 4006;
```

Comment:
Could you add the support for the environment variable `PORT` please ? It would help me to deploy this on heroku for the workshop